File: packages/shared/src/map.ts
import type { TileMap } from './types';

/**
 * Parse a JSON string into a TileMap, validating required fields.
 */
export function parseMap(json: string): TileMap {
  const data = JSON.parse(json);

  if (typeof data.width !== 'number') {
    throw new Error('Map missing required field: width');
  }
  if (typeof data.height !== 'number') {
    throw new Error('Map missing required field: height');
  }
  if (!Array.isArray(data.tiles)) {
    throw new Error('Map missing required field: tiles');
  }
  if (data.tiles.length !== data.width * data.height) {
    throw new Error(
      `Map tiles length (${data.tiles.length}) does not match width*height (${data.width * data.height})`,
    );
  }

  return {
    width: data.width,
    height: data.height,
    tiles: data.tiles,
  };
}

/**
 * Generate a test arena with walls, corridors, and open spaces.
 *
 * Layout:
 * - 2-tile thick border walls
 * - Central open area (~40x40) for dogfighting
 * - 4 quadrant rooms connected by corridors
 * - Wall pillars/obstacles scattered in open areas
 */
export function generateTestArena(width: number, height: number): TileMap {
  const tiles = new Array(width * height).fill(0);

  function setTile(x: number, y: number, value: number): void {
    if (x >= 0 && x < width && y >= 0 && y < height) {
      tiles[y * width + x] = value;
    }
  }

  function fillRect(
    x1: number,
    y1: number,
    x2: number,
    y2: number,
    value: number,
  ): void {
    for (let y = y1; y <= y2; y++) {
      for (let x = x1; x <= x2; x++) {
        setTile(x, y, value);
      }
    }
  }

  // 2-tile thick border walls
  fillRect(0, 0, width - 1, 1, 1); // top
  fillRect(0, height - 2, width - 1, height - 1, 1); // bottom
  fillRect(0, 0, 1, height - 1, 1); // left
  fillRect(width - 2, 0, width - 1, height - 1, 1); // right

  const cx = Math.floor(width / 2);
  const cy = Math.floor(height / 2);

  // Cross-shaped walls dividing map into quadrants
  // Horizontal wall with gaps (corridors)
  for (let x = 10; x < width - 10; x++) {
    if (
      Math.abs(x - cx) > 5 && // gap at center
      Math.abs(x - Math.floor(cx / 2)) > 3 && // gap at quarter
      Math.abs(x - Math.floor((cx * 3) / 2)) > 3 // gap at three-quarter
    ) {
      setTile(x, cy, 1);
      setTile(x, cy - 1, 1);
    }
  }

  // Vertical wall with gaps
  for (let y = 10; y < height - 10; y++) {
    if (
      Math.abs(y - cy) > 5 && // gap at center
      Math.abs(y - Math.floor(cy / 2)) > 3 && // gap at quarter
      Math.abs(y - Math.floor((cy * 3) / 2)) > 3 // gap at three-quarter
    ) {
      setTile(cx, y, 1);
      setTile(cx - 1, y, 1);
    }
  }

  // Corner room walls in each quadrant
  const quadrants = [
    { x: Math.floor(cx / 2), y: Math.floor(cy / 2) },
    { x: Math.floor((cx * 3) / 2), y: Math.floor(cy / 2) },
    { x: Math.floor(cx / 2), y: Math.floor((cy * 3) / 2) },
    { x: Math.floor((cx * 3) / 2), y: Math.floor((cy * 3) / 2) },
  ];

  for (const q of quadrants) {
    // L-shaped wall structures
    fillRect(q.x - 8, q.y - 8, q.x - 8, q.y - 3, 1); // vertical segment
    fillRect(q.x - 8, q.y - 8, q.x - 3, q.y - 8, 1); // horizontal segment

    fillRect(q.x + 3, q.y + 3, q.x + 8, q.y + 3, 1); // horizontal segment
    fillRect(q.x + 8, q.y + 3, q.x + 8, q.y + 8, 1); // vertical segment

    // Pillar obstacles (2x2)
    fillRect(q.x - 1, q.y - 1, q.x, q.y, 1);
  }

  // Additional corridor walls for complexity
  // Top-left diagonal-like structure
  fillRect(20, 20, 25, 20, 1);
  fillRect(25, 20, 25, 25, 1);

  // Bottom-right structure
  fillRect(width - 26, height - 26, width - 21, height - 26, 1);
  fillRect(width - 26, height - 26, width - 26, height - 21, 1);

  return { width, height, tiles };
}

File: packages/server-rs/src/game/map.rs
/// Tile map: 0=empty, 1=wall. Row-major flat array.
#[derive(Debug, Clone)]
pub struct TileMap {
    pub width: u32,
    pub height: u32,
    pub tiles: Vec<u32>,
}

impl TileMap {
    pub fn is_wall_at(&self, tx: i32, ty: i32) -> bool {
        if tx < 0 || tx >= self.width as i32 || ty < 0 || ty >= self.height as i32 {
            return true;
        }
        self.tiles[(ty as u32 * self.width + tx as u32) as usize] != 0
    }

    /// Check if a circle (AABB approximation) collides with any wall tile.
    pub fn is_colliding(&self, x: f32, y: f32, radius: f32) -> bool {
        let min_tx = (x - radius).floor() as i32;
        let max_tx = (x + radius).floor() as i32;
        let min_ty = (y - radius).floor() as i32;
        let max_ty = (y + radius).floor() as i32;

        for ty in min_ty..=max_ty {
            for tx in min_tx..=max_tx {
                if self.is_wall_at(tx, ty) {
                    let ship_left = x - radius;
                    let ship_right = x + radius;
                    let ship_top = y - radius;
                    let ship_bottom = y + radius;

                    let tile_left = tx as f32;
                    let tile_right = tx as f32 + 1.0;
                    let tile_top = ty as f32;
                    let tile_bottom = ty as f32 + 1.0;

                    if ship_right > tile_left
                        && ship_left < tile_right
                        && ship_bottom > tile_top
                        && ship_top < tile_bottom
                    {
                        return true;
                    }
                }
            }
        }
        false
    }
}

/// Apply wall collision using axis-separated approach (X then Y).
pub fn apply_wall_collision(
    x: &mut f32,
    y: &mut f32,
    vx: &mut f32,
    vy: &mut f32,
    dt: f32,
    map: &TileMap,
    radius: f32,
    bounce_factor: f32,
) {
    // X axis
    let prev_x = *x;
    *x += *vx * dt;
    if map.is_colliding(*x, *y, radius) {
        *x = prev_x;
        *vx *= -bounce_factor;
        *vy *= bounce_factor;
    }

    // Y axis
    let prev_y = *y;
    *y += *vy * dt;
    if map.is_colliding(*x, *y, radius) {
        *y = prev_y;
        *vy *= -bounce_factor;
        *vx *= bounce_factor;
    }
}

fn set_tile(tiles: &mut [u32], width: u32, height: u32, x: i32, y: i32, value: u32) {
    if x >= 0 && (x as u32) < width && y >= 0 && (y as u32) < height {
        tiles[(y as u32 * width + x as u32) as usize] = value;
    }
}

fn fill_rect(tiles: &mut [u32], width: u32, height: u32, x1: i32, y1: i32, x2: i32, y2: i32, value: u32) {
    for y in y1..=y2 {
        for x in x1..=x2 {
            set_tile(tiles, width, height, x, y, value);
        }
    }
}

/// Generate a test arena matching the TypeScript generateTestArena.
pub fn generate_test_arena(width: u32, height: u32) -> TileMap {
    let mut tiles = vec![0u32; (width * height) as usize];

    // 2-tile thick border walls
    fill_rect(&mut tiles, width, height, 0, 0, width as i32 - 1, 1, 1);
    fill_rect(&mut tiles, width, height, 0, height as i32 - 2, width as i32 - 1, height as i32 - 1, 1);
    fill_rect(&mut tiles, width, height, 0, 0, 1, height as i32 - 1, 1);
    fill_rect(&mut tiles, width, height, width as i32 - 2, 0, width as i32 - 1, height as i32 - 1, 1);

    let cx = (width / 2) as i32;
    let cy = (height / 2) as i32;

    // Cross-shaped walls dividing map into quadrants
    for x in 10..(width as i32 - 10) {
        if (x - cx).abs() > 5
            && (x - cx / 2).abs() > 3
            && (x - (cx * 3) / 2).abs() > 3
        {
            set_tile(&mut tiles, width, height, x, cy, 1);
            set_tile(&mut tiles, width, height, x, cy - 1, 1);
        }
    }

    for y in 10..(height as i32 - 10) {
        if (y - cy).abs() > 5
            && (y - cy / 2).abs() > 3
            && (y - (cy * 3) / 2).abs() > 3
        {
            set_tile(&mut tiles, width, height, cx, y, 1);
            set_tile(&mut tiles, width, height, cx - 1, y, 1);
        }
    }

    // Corner room walls in each quadrant
    let quadrants = [
        (cx / 2, cy / 2),
        ((cx * 3) / 2, cy / 2),
        (cx / 2, (cy * 3) / 2),
        ((cx * 3) / 2, (cy * 3) / 2),
    ];

    for &(qx, qy) in &quadrants {
        // L-shaped wall structures
        fill_rect(&mut tiles, width, height, qx - 8, qy - 8, qx - 8, qy - 3, 1);
        fill_rect(&mut tiles, width, height, qx - 8, qy - 8, qx - 3, qy - 8, 1);
        fill_rect(&mut tiles, width, height, qx + 3, qy + 3, qx + 8, qy + 3, 1);
        fill_rect(&mut tiles, width, height, qx + 8, qy + 3, qx + 8, qy + 8, 1);

        // Pillar obstacles (2x2)
        fill_rect(&mut tiles, width, height, qx - 1, qy - 1, qx, qy, 1);
    }

    // Additional corridor walls
    fill_rect(&mut tiles, width, height, 20, 20, 25, 20, 1);
    fill_rect(&mut tiles, width, height, 25, 20, 25, 25, 1);
    fill_rect(&mut tiles, width, height, width as i32 - 26, height as i32 - 26, width as i32 - 21, height as i32 - 26, 1);
    fill_rect(&mut tiles, width, height, width as i32 - 26, height as i32 - 26, width as i32 - 26, height as i32 - 21, 1);

    TileMap { width, height, tiles }
}
